### ISSUE TITLE
Add leaky ReLU neural layer

### ISSUE THREAD

**Issue by lisitsyn** [shogun-toolbox/shogun#2699]
Good for [Deep learning](https://github.com/shogun-toolbox/shogun/wiki/GSoC_2015_project_deep_learning) applicants

**Comment by sanuj** [comment]
@lisitsyn I am interested in deep learning. I have read about neural nets but never really implemented them. Of what I know, rectifier is an activation function `max(0,x)` and leaky ReLU is when it has some small gradient for x<0. I have used the sigmoid function before but not this(but this should not be a problem I guess). I also had a look at the two papers that you mentioned (text understanding and image classification) and found out that CNNs are rather important for computer vision :) I have never used them but I will definitely read up about them. I know about word2vec that was mentioned in the text understanding paper from the kaggle beginner problem (sentimental analysis) :) So I am relatively new to deep learning but eager to learn.
I chose this because it seemed easiest of the three. I'll implement the papers after this if no-one is doing that.
We already have a `CNeuralRectifiedLinearLayer` but I guess we will need a separate layer for leaky ReLU. Anything that I should know about before I start?

**Comment by sanuj** [comment]
Should we go by adding a new class which inherits `NeuralLinearLayer`? Something like -
`class CNeuralLeakyRectifiedLinearLayer : public CNeuralLinearLayer`

**Comment by lisitsyn** [comment]
@sanuj sorry missed this message somehow.

It is true CNNs are useful for computer vision but this recent paper I mentioned uses it for text understanding quite successfully so it is worth investigating.
 
As for your mention of sigmoid thing - ReLUs are usually better due to no vanishing gradient effect (it is rather impossible to train a deep net with sigmoids) and the speed of its evaluation. 

Feel free to submit patches for these tasks on neural nets if you are interested!

**Comment by sanuj** [comment]
Cool @lisitsyn :)

**Comment by vrishank97** [comment]
Can I take this up as my first contribution here?

**Comment by vigsterkr** [comment]
@vrishank97 sure

**Comment by vrishank97** [comment]
I think this issue has already been taken care of - https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/neuralnets/NeuralLeakyRectifiedLinearLayer.cpp

**Comment by vigsterkr** [comment]
oh, then this is obsolete! :)
